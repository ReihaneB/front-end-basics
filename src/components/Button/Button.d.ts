import React from 'react';

interface ButtonProps {
  onClick: () => void;
  children: React.ReactNode;
  disabled?: boolean;
  disableDebounce?: boolean;
  autoFocus?: boolean;
  forwardedRef?: React.RefObject<HTMLButtonElement> | null;
}
