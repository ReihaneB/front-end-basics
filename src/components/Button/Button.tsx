import { memo } from 'react';

import { useClickable } from '@/hooks/useClickable/useClickable';
import { useFocusable } from '@/hooks/useFocusable/useFocusable';

import type { ButtonProps } from './Button.d';

function Button({
  onClick,
  children,
  disabled = false,
  disableDebounce = false,
  autoFocus = false,
  forwardedRef = null,
  ...rest
}: ButtonProps) {
  const { eventHandlers } = useClickable({
    onClick,
    disabled,
    disableDebounce,
  });

  const { ref: elementRef } = useFocusable(forwardedRef, { autoFocus });

  return (
    <button
      type="button"
      ref={elementRef}
      disabled={disabled}
      data-testid="button-component"
      {...eventHandlers}
      {...rest}
    >
      {children}
    </button>
  );
}

export default memo(Button);
