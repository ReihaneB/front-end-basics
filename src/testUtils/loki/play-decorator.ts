/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-expect-error
import createAsyncCallback from '@loki/create-async-callback';
// @ts-expect-error
import isLokiRunning from '@loki/is-loki-running';
import type {PlayFunction} from '@storybook/csf';
// @ts-expect-error
import type {ReactFramework} from '@storybook/react';

async function waitForDocumentLoaded(): Promise<void> {
  if (document.readyState === 'loading') {
    return new Promise(resolve => {
      document.addEventListener('DOMContentLoaded', () => resolve());
    });
  }

  return Promise.resolve();
}

function enablePointerEvents(): void {
  const styleElement = document.createElement('style');
  document.documentElement.appendChild(styleElement);
  styleElement.sheet?.insertRule('* {pointer-events: auto !important}');
}

export function lokiPlayDecorator(
  target: PlayFunction<ReactFramework>
): PlayFunction<ReactFramework> {
  return async context => {
    const lokiRunning = isLokiRunning();
    const asyncCallback = createAsyncCallback();

    if (lokiRunning) {
      await waitForDocumentLoaded();
      enablePointerEvents();
    }

    try {
      await target(context);
    } finally {
      asyncCallback();
    }
  };
}
