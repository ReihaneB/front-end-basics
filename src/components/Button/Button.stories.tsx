import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, waitFor, within } from '@storybook/test';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';

import Button from './Button';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    onClick: {
      action: 'clicked',
    },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

const defaultArgs = {
  children: 'Click me',
  disabled: false,
  disableDebounce: false,
  autoFocus: false,
  forwardedRef: null,
};

export const Playground: Story = {
  args: {
    ...defaultArgs,
  },
};

export const Disabled: Story = {
  args: {
    ...defaultArgs,
    disabled: true,
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Click the button 2 times', async () => {
        await userEvent.click(canvas.getByTestId('button-component'));
        await userEvent.click(canvas.getByTestId('button-component'));
      });

      await waitFor(() => expect(args.onClick).toHaveBeenCalledTimes(0));
    }
  ),
};

export const AutoFocus: Story = {
  args: {
    ...defaultArgs,
    autoFocus: true,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await waitFor(() => expect(canvas.getByTestId('button-component')).toHaveFocus());
    }
  ),
};

export const ClickMultipleTimes: Story = {
  name: 'Click multiple times but only call once',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Click the button 2 times', async () => {
        await userEvent.click(canvas.getByTestId('button-component'));
        await userEvent.click(canvas.getByTestId('button-component'));
      });

      await waitFor(() => expect(args.onClick).toHaveBeenCalledTimes(1));
    }
  ),
};

export const DisableDebounce: Story = {
  args: {
    ...defaultArgs,
    disableDebounce: true,
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Click the button 2 times', async () => {
        await userEvent.click(canvas.getByTestId('button-component'));
        await userEvent.click(canvas.getByTestId('button-component'));
      });

      await waitFor(() => expect(args.onClick).toHaveBeenCalledTimes(2));
    }
  ),
};

export const PressEnter: Story = {
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Press the button 2 times', async () => {
        await userEvent.type(canvas.getByTestId('button-component'), '{Enter}');
        await userEvent.type(canvas.getByTestId('button-component'), '{Enter}');
      });

      await waitFor(() => expect(args.onClick).toHaveBeenCalledTimes(1));
    }
  ),
};

export const PressSpace: Story = {
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Press the button 2 times', async () => {
        await userEvent.type(canvas.getByTestId('button-component'), '{Space}');
        await userEvent.type(canvas.getByTestId('button-component'), '{Space}');
      });

      await waitFor(() => expect(args.onClick).toHaveBeenCalledTimes(1));
    }
  ),
};

export const FocusOnClick: Story = {
  name: 'Focused when clicked',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await userEvent.click(canvas.getByTestId('button-component'));

      await waitFor(() => expect(canvas.getByTestId('button-component')).toHaveFocus());
    }
  ),
};

export const FocusOnTab: Story = {
  name: 'Focused when tabbed',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await userEvent.tab();

      await waitFor(() => expect(canvas.getByTestId('button-component')).toHaveFocus());
    }
  ),
};
