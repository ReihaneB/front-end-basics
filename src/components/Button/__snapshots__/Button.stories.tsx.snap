// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Button AutoFocus play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button ClickMultipleTimes play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button DisableDebounce play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button Disabled play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          disabled
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button FocusOnClick play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button FocusOnTab play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button Playground smoke-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button PressEnter play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;

exports[`Components/Button PressSpace play-test 1`] = `
<main class="inter-normal">
  <button type="button"
          data-testid="button-component"
  >
    Click me
  </button>
</main>
`;
